Parse version/service and remove generateVersion()

chanseokoh opened the pull request:
See #655.

Note that this should not be merged until the internal #120830848 makes to public.

This PR
- removes generateVersion()
- parses the JSON output from `gcloud app deploy` to get version and service info (this is a bit ugly)

Unlike JavaScript, parsing JSON in Java was a bit of a pain. Because we don't know the exact format/type of the JSON output and it may change in the future, I took an approach of manually going over each JSON element to try best to extract the version/service info. I wonder if there is a nicer way.

An example output:
```json
        {
          "configs": [],
          "versions": [
            {
              "id": "20160429t112518",
              "last_deployed_time": null,
              "project": "springboot-maven-project",
              "service": "default",
              "traffic_split": null,
              "version": null
            }
          ]
        }
```

Probably, I should move `parseDeployOutputToModule()` to the `AppEngineUtil` class and write a test.

Tried basic testing, and undeploying (stopping) works correctly.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -247,15 +241,72 @@ public void run(@NotNull ProgressIndicator indicator) {
+      reader.beginObject(); // Top-level is a single object.
+
+      // Look for "versions" in the top-level object.
+      while (reader.hasNext()) {
```
Might be cleaner if instead you create a separate object respresenting the deploy output that you can marshal the json to.

chanseokoh replied:
That would be nice, actually. My only concern is the possibility of future changes to the JSON output. Will that kind of marshallable object resilient/flexible enough to adapt changes in JSON. I hadn't tested it, but it feels like if anything is out of the type slightly, Gson will fail.

Related to that, it is not clear, for example, if `"traffic_split": null` will have a string type or may have an array.

etanshaul replied:
We don't need the traffic split info right? So that one can be ignored. 

Wouldn't you're current setup also fail if they change the output? E.g. you expect a versions at the top level and then look inside for specific attributes; if this changes then you wouldn't be able to extract the output any more.

chanseokoh replied:
I see. So I can just omit fields that I am interested in the marshallable object, right? That will certainly do.

etanshaul replied:
I'm not sure if you omit them or if you control it via gson annotations in your "domain" object that you create to represent the deploy output.

chanseokoh commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -247,15 +241,72 @@ public void run(@NotNull ProgressIndicator indicator) {
+                reader.skipValue(); // Skip a value if not "id".
+              }
+            }
+            reader.endObject();
```
I feel like this will fail if "versions" contain multiple elements. I will need to fix this.

chanseokoh commented:
- Did an overhaul to use a JSON-matching class for deserialization as @etanshaul suggested. Now this is much better. It resulted in a few FindBugs warnings of accessing seemingly uninitialized values of this JSON-matching class. I will think about what I can do with the warnings.

- Also moved the static JSON-parsing function to `AppEngineUtil`.

- Added parsing tests.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/util/AppEngineUtil.java:
```diff
@@ -16,34 +16,64 @@
+  }
+
+  // Holds de-serialized JSON output of gcloud app deploy.
+  private static class DeployOutput {
```
Having these objects as static classes in AppEngineUtil doesn't feel fight. I would suggest extracting them out into their own class (you can still keep them a single file if you prefer).

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/util/AppEngineUtil.java:
```diff
@@ -16,34 +16,64 @@
 public class AppEngineUtil {
 
+  // Just to return two strings.
+  public static class VersionService {
```
I don't think you need this. If you like, you could make a helper getters in your DeploymentOutput class to give you these values.

chanseokoh replied:
Yeah, I want to avoid VersionService too. I will add getters in DeployOutput.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/util/AppEngineUtil.java:
```diff
@@ -16,34 +16,64 @@
+
+  // Holds de-serialized JSON output of gcloud app deploy.
+  private static class DeployOutput {
+    private static class versionElement {
```
Capitalization. How about just `Version` instead of VersionElement? Also the nesting here feels a bit odd.

etanshaul replied:
The nesting might be ok actually. I'd just consider if AppEngineUtil is the right place for this.

chanseokoh replied:
Yeah, the parsing function is too-specific. Where could be a good place for this? Going back?

etanshaul replied:
I don't know if it makes sense for it to be static since its related to an instance of the deployment. I think it would be fine to put your updated gson logic back into a private method of the DeployAction class.

chanseokoh commented:
Come to think of it, a fix to this #655 is tricky to roll out. Unless a user upgrades our plugin and gcloud at the same time, things will not work. We are not in an ideal world, so probably this PR needs to attempt to parse JSON first with the old format and later with the new format.

etanshaul commented:
We already face this situation with the cloud sdk version. If the deployment fails (for whatever reason) we print out a message suggesting to the user to upgrade to the latest version. See AppEngineDeployAction#processTerminated

chanseokoh commented:
Or, since everything is BETA, we could just not care about breakages caused by using wrong gcloud.

chanseokoh commented:
Okay, I will do the same thing to print out an upgrade message.

etanshaul commented:
You shouldn't have to. Its already handled. Just make sure the deployment fails if the output isn't what you expected.

chanseokoh commented:
I noticed that this parsing routine is executed only when you click 'Undeploy'. Deployment can go fine but parsing can fail.

etanshaul commented:
Ah you're right. Its a little tricky then because this parsing occurs outside the context of a command line process.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -184,11 +182,9 @@ public void processTerminated(final ProcessEvent event) {
 
   private class DeploymentRuntimeImpl extends DeploymentRuntime {
```
could this be broken out and named something like AppEngineDeploymentRuntime?

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -221,18 +217,19 @@ public void run() {
-      String moduleToStop;
+      DeployOutput deployOutput;
       try {
-        moduleToStop = parseDeployOutputToModule(deploymentOutput);
```
what do you think about parsing this in the constructor?

etanshaul replied:
Not that performance is really a concern here but just to explain the original reason I did this: having it in the stop() method lazily does the json parsing (doesn't happen until the user actually tries to undeploy which may never happen in the first place).

patflynn replied:
Yeah that's reasonable.  But if our parsing of the deployment result is
broken, I would prefer that we broke immediately, rather than lazily during
stop.  For one we'd be much more likely to notice the breakage.


On Mon, May 2, 2016 at 10:51 AM, Etan Shaul <notifications@github.com>
wrote:

> In
> google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployAction.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/660#discussion_r61748614>
> :
>
> >        try {
> > -        moduleToStop = parseDeployOutputToModule(deploymentOutput);
>
> Not that performance is really a concern here but just to explain the
> original reason I did this: having it in the stop() method lazily does the
> json parsing (doesn't happen until the user actually tries to undeploy
> which may never happen in the first place).
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/660/files/ddd3d9a477599546aacb5636136c10cad7eebe79#r61748614>
>

etanshaul replied:
Yea. That sounds good to me.

chanseokoh replied:
SGTM. I'll move it to the constructor. The constructor will not throw any exception.

chanseokoh commented:
Notable changes since last commit:
- Refactored DeploymentRuntimeImpl to be a top-level class.
- Moved parsing to the constructor.
- If the JSON parsing fails, the 'Undeploy' menu item will be disabled (grayed out).
- Added one more test case for parsing.

chanseokoh commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployActionTest.java:
```diff
@@ -82,16 +80,12 @@ public void testCreateApplicationDefaultCredentials() throws Exception {
-  @After
-  public void deleteCredentialFile() {
     credentialFile.delete();
   }
```
This change is necessary because `@After` method is called for each `@Test`.

chanseokoh commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java:
```diff
@@ -0,0 +1,181 @@
+  @Override
+  public void undeploy(@NotNull final UndeploymentTaskCallback callback) {
+    if (deployOutput == null) {
+      callback.errorOccurred(GctBundle.message("appengine.stop.modules.version.execution.error"));
```
This check is probably not necessary since 'Undeploy' will be disabled and users will not be able to get into this function if `deployOutput == null`. (See the above `isUndeploySupported()` function.) I left the check here just in case.

chanseokoh commented:
Tested this with both on the public gcloud (old format) and the internal gcloud (new format). Undeploy works correctly with the new JSON format.

I still need to address issues reported by FindBugs that is confused by Gson's magical field initialization (i.e., de-serialization). I will have to use FindBugs annotations.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java:
```diff
@@ -0,0 +1,184 @@
+    } catch (JsonParseException e) {
+      loggingHandler.print(
+          GctBundle.message("appengine.deployment.version.extraction.failure") + "\n");
+      Logger.getInstance(AppEngineDeploymentRuntimeImpl.class).warn(
```
Loggers should be statically initialized at the top of the class. 

Also I think we probably want to have this kind of error reported right?

chanseokoh replied:
I'll move the logger.

If you mean reporting as a crash report, in the long run that would be the right direction. But we will have a lot of such (false) reports when users upgrade our plugin while they don't upgrade gcloud (unless we put a special logic to identify and exclude errors caused by the old gcloud format).

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java:
```diff
@@ -0,0 +1,184 @@
+/**
+ * Takes care of undeploy and stop of App Engine applications.
+ */
+class AppEngineDeploymentRuntimeImpl extends DeploymentRuntime {
```
you don't need to call this ..Impl anymore.

chanseokoh replied:
Interesting! Anymore... means, like, since Java 1.7? Or like, because it is no longer a nested function?

patflynn replied:
no, what I meant was that the ..Impl  in the class name isn't necessary now that you've qualified it as an AppEngineDeploymentRuntime.  You can just call it that.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java:
```diff
@@ -0,0 +1,184 @@
+          GctBundle.message("appengine.deployment.version.extraction.failure") + "\n");
+      Logger.getInstance(AppEngineDeploymentRuntimeImpl.class).warn(
+          "Could not retrieve service/version info of deployed application", e);
+      deployOutput = null;
```
if parsing failed, this object seems to be fundamentally useless. I'd rather we didn't construct it at all. What do you think?

etanshaul replied:
This is a little subtle:

If the deploy succeeds, but the parsing fails, then the only way to indicate a successful deploy is by constructing that class (and passing it to `callback.succeeded`). 

This _shouldn't_ happen, but I can imagine a case where the deploy succeeded, but gcloud returns an unexpected deploy output -> in this case I think we'd want to still indicate a successful deploy.

etanshaul replied:
One improvement would be to move the parsing to the DeployAction. Then if its successful construct the Runtime with undeploy enabled (and disabled if not). 

We still have that useless Runtime but thats the pattern that IJ uses: even if undeploy was not supported at all we still have to construct it and override `isUndeploySupported` because of the callback api.

chanseokoh replied:
I agree on that JSON parsing failure is still a successful deployment, because it is true that the app is actually deployment and running in this case.

As @etanshaul said, even if we move the parsing to DeployAction, we will still have to construct (possibly another) DeploymentRuntime anyway to override `isUndeploySupported` to return false. In this sense, I think it is not a bad idea to do parsing in this constructor. Ideas?

patflynn replied:
I think etan's suggestion to do the parsing in the action and construct an
appropriate runtime based off of the result of that could work.

On Tue, May 3, 2016 at 1:18 PM, Chanseok Oh <notifications@github.com>
wrote:

> In
> google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/660#discussion_r61919422>
> :
>
> > +      @NotNull Project project,
> > +      @NotNull AppEngineHelper appEngineHelper,
> > +      @NotNull LoggingHandler loggingHandler,
> > +      @NotNull String deploymentOutput) {
> > +    this.project = project;
> > +    this.appEngineHelper = appEngineHelper;
> > +    this.loggingHandler = loggingHandler;
> > +
> > +    try {
> > +      deployOutput = parseDeployOutputToService(deploymentOutput);
> > +    } catch (JsonParseException e) {
> > +      loggingHandler.print(
> > +          GctBundle.message("appengine.deployment.version.extraction.failure") + "\n");
> > +      Logger.getInstance(AppEngineDeploymentRuntimeImpl.class).warn(
> > +          "Could not retrieve service/version info of deployed application", e);
> > +      deployOutput = null;
>
> I agree on that JSON parsing failure is still a successful deployment,
> because it is true that the app is actually deployment and running in this
> case.
>
> As @etanshaul <https://github.com/etanshaul> said, even if we move the
> parsing to DeployAction, we will still have to construct (possibly another)
> DeploymentRuntime anyway to override isUndeploySupported to return false.
> In this sense, I think it is not a bad idea to do parsing in this
> constructor. Ideas?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/660/files/5f2176fab1a2f91586c3eb81b376971583718b4a#r61919422>
>

etanshaul replied:
@chanseokoh just noting that there's no need to construct a different Runtime depending on if isUndeploySupported. You can just pass in a boolean to the constructor.

As far as I can see, you can still do the parsing in the deploy action.

chanseokoh replied:
@etanshaul Aha, nice catch.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java:
```diff
@@ -0,0 +1,184 @@
+  }
+
+  @VisibleForTesting
+  static DeployOutput parseDeployOutputToService(String jsonOutput) throws JsonParseException {
```
drop the ..toService.  parseDeployOutput(jsonOutput). It's cleaner.

chanseokoh replied:
Done.

chanseokoh commented:
- Parsing is done in `AppEngineDeployAction`. Moved tests as well.
- `DeployOutput` Gson-object methods can now return null.
- Renamed to `AppEngineDeploymentRuntime`.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -167,7 +155,20 @@ public void onTextAvailable(ProcessEvent event, Key outputType) {
+          try {
+            deployOutput = parseDeployOutput(deploymentOutput.toString());
+          } catch (JsonParseException e) {
+            getLoggingHandler().print(
```
you can use consoleLogLn here instead. Also, did we settle on not being notified on this kind of failure (I don't think we get a report on logger.warn).  Otherwise, lgtm.

patflynn replied:
I think it's OK to error here.  You could conceivably explain the error and suggest their might be an incompatibility with gcloud that could be fixed with a component update.

chanseokoh replied:
For both `consoleLogLn()` and `logger.error()`, they actually don't need text localization. However, not using `GctBundle.message()` results in embedding a very long string in the code such as below. It will also make a bit difficult to maintain consistent messaging if it's outside the resrouce bundle.

"Failed to parse service/version info from the ''gcloud'' deployment output. \nPlease make sure that you are using the latest version of the Google Cloud SDK.\nRun ''gcloud components update'' to update the SDK. (See: https://cloud.google.com/sdk/gcloud/reference/components/update.)"

Suggestions?

patflynn replied:
I think what you have there is fine.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntime.java:
```diff
@@ -0,0 +1,112 @@
+
+  @Override
+  public void undeploy(@NotNull final UndeploymentTaskCallback callback) {
+    if (!isUndeploySupported()) {
```
oh, you can get rid of this block since it can never be executed.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntime.java:
```diff
@@ -0,0 +1,112 @@
+
+  @Override
+  public boolean isUndeploySupported() {
+    return service != null && version != null;
```
should this be OR and not AND?

chanseokoh replied:
Both of them are required to undeploy an app, right?

etanshaul replied:
Yes, for our purposes we want both. I think you had it correct. But see my other comment about the possible bug.

patflynn replied:
ah I had this misread.  My brain expects to see !(service == null || version == null). This is fine though.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -167,7 +155,20 @@ public void onTextAvailable(ProcessEvent event, Key outputType) {
+
+          callback.succeeded(
+              new AppEngineDeploymentRuntime(project, appEngineHelper, getLoggingHandler(),
+                                             deployOutput.getService(), deployOutput.getVersion()));
```
wouldn't this throw an NPE if the parsing fails since deployOutput will be null?

chanseokoh replied:
Wow.

chanseokoh commented:
- `consoleLogLn`, and print the message whenever it fails to get version and service.
- `logger.error` instead of `logger.warn`. This allows user bug reporting.
- No unnecessary check at the beginning of `AppEngineDeploymentRuntime.undeploy()` as @etanshaul suggested.

When using the current `gcloud`:
![old_gcloud](https://cloud.githubusercontent.com/assets/10523105/14996706/556ca2ae-1149-11e6-8541-8290078f1ee2.png)
![old_gcloud_events](https://cloud.githubusercontent.com/assets/10523105/14996712/5c1de9fa-1149-11e6-9e01-265b21b3300d.png)

When using the new `gcloud`:![new_gcloud](https://cloud.githubusercontent.com/assets/10523105/14996718/5fa8e872-1149-11e6-8d46-292081cb0a10.png)

patflynn commented:
If all comments are addressed: :ship:
